### ISSUE TITLE
Use msgpack for signature serialization/deserialization

### ISSUE THREAD

**Issue by BoPeng** [vatlab/SoS#1174]
Currently we use compressed `pickle` for the serialization and deserialization of signatures, which are saved to sqlite databases. We should evaluate the possibility of using msgpack because it is said to be small and efficient.

**Comment by gaow** [comment]
+1 for that. Are you saying replacing the sqlite databases with msgpack? Is our database a simple key-value database? I've been using msgpack for as a drop-in replacement for JSON. Not sure if it is possible to load a key without loading the entire msgpack file though, that seems the biggest limiitation?

**Comment by BoPeng** [comment]
No, sqlite databases will continue to be used but the way signatures are serialized can be improved. Although currently signature serialization does not appear to be a bottleneck due to the inefficiently of other parts of SoS, this is something we need to visit sooner or later to improve the overall efficiency of SoS.

**Comment by BoPeng** [comment]
Since I mostly handles small number of large jobs, it would be helpful for you to attach with this ticket (or PM me) a decent sized signature database (`.sos/*.db`) and I can use separate code to read/write it to compare the performance.

**Comment by gaow** [comment]
Oh so you envisage it is a replacement of `pickle`? Does it support all the data types we need in signature as `pickle` does? This would also be a drop-in replacement of pickle right?

How about [this database for test](http://shiny.stephenslab.uchicago.edu/gaow/signatures_34K.tar.gz)

it is from one of the earlier SoS versions please let me know if it suffices. I can certainly come up with more of it. It's just that due to recent version changes I've been proactively removing my `.sos` folders on the file system but I can always geenrate more of those easily.

**Comment by BoPeng** [comment]
I do not know yet.

I have been wary about allowing the pickle/unpickle of arbitrary pickleable datatypes. We will have to restrict the datatypes if we switch to msgpack but that does not have to be a bad thing, especially if it makes SoS more robust and efficient.

**Comment by gaow** [comment]
I see. I guess maybe we first evaluate the performance gain to see if it is worth switching to it first, then worry about other performance bottleneck and making restrictions to improve efficiency and robustness.

**Comment by BoPeng** [comment]
Note the use of msgpack by dask

https://distributed.dask.org/en/latest/protocol.html#

Dask is pretty efficiency so their choice should be the result of numerous tests.
